x/dymns/keeper/hooks.go

```diff
@@ -0,0 +1,365 @@
+				return nil
+			}
+
+			if err := e.CompleteDymNameSellOrder(ctx, so.AssetId); err != nil {
+				logger.Error("failed to complete sell order", "asset-id", so.AssetId, "error", err)
+				return err
+			}
+
+			return nil
+		})
+
+		if errApplyStateChange == nil {
+			aSoe.Remove(so.AssetId)
+		}
+	}
+
+	if err := e.SetActiveSellOrdersExpiration(ctx, aSoe, dymnstypes.TypeName); err != nil {
+		e.Keeper.Logger(ctx).Error("failed to update active SO expiry", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// processActiveAliasSellOrders process the finished Alias Sell-Orders.
+// Sell-Order will be deleted. If the Sell-Order has a winner, the Alias linking will be updated.
+func (e epochHooks) processActiveAliasSellOrders(ctx sdk.Context, logger log.Logger) error {
+	finishedSOs, aSoe := e.getFinishedSellOrders(ctx, dymnstypes.TypeAlias, logger)
+
+	if len(finishedSOs) < 1 || len(aSoe.Records) < 1 {
+		return nil
+	}
+
+	logger.Info("processing finished SOs", "count", len(finishedSOs))
+
+	prohibitedToTradeAliases := e.GetAllAliasAndChainIdInParams(ctx)
+
+	for _, so := range finishedSOs {
+		// each order should be processed in a branched context, if error, discard the state change
+		// and process next order, to prevent chain reaction when an individual order failed to process
+		errApplyStateChange := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			if so.HighestBid == nil {
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeAlias)
+				return nil
+			}
+
+			/**
+			For the Sell-Orders which the assets are prohibited to trade,
+			the Sell-Order will be force cancelled and the bids will be refunded.
+
+			Why some aliases are prohibited to trade? And what are they?
+			In module params, there is a list of alias mapping for some external well-known chains.
+			So those aliases are considered as reserved for the external chains,
+			therefor trading is not allowed.
+
+			Why can someone own a prohibited alias?
+			An alias can be bought before the reservation was made.
+			But when the alias becomes reserved for the external well-known chains,
+			the alias will be prohibited to trade.
+
+			Why can someone place a Sell-Order for the prohibited alias?
+			When a Sell-Order created before the reservation was made.
+			*/
+			if _, forceCancel := prohibitedToTradeAliases[so.AssetId]; forceCancel {
+				// Sell-Order will be force cancelled and refund bids if any,
+				// when the alias is prohibited to trade
+				if err := e.RefundBid(ctx, *so.HighestBid, dymnstypes.TypeAlias); err != nil {
+					logger.Error("failed to refund bid for a force-to-cancel sell order", "asset-id", so.AssetId, "error", err)
+					return err
+				}
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeAlias)
+				return nil
+			}
+
+			if err := e.CompleteAliasSellOrder(ctx, so.AssetId); err != nil {
+				logger.Error("failed to complete sell order", "asset-id", so.AssetId, "error", err)
+				return err
+			}
+
+			return nil
```

Comment:
why do we set active sell order expiration in that case?

Reply:
That's a list of expiration date, instead of iterating store, I persist every expiration into one record.